**Nhóm 2:**

  + Nguyễn Trần Ngọc Ty - 23521758

  + Nguyễn Huỳnh Minh Phú - 23521183

  + Hà Xuân Hoàng - 23520516

# 1. Thư viện

In [ ]:
import pandas as pd
import numpy as np
import warnings
from datetime import datetime, timedelta
import re
warnings.filterwarnings("ignore")

Kết nối đường dẫn đến google drive chứa dữ liệu

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# 2. Đọc dữ liệu

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/DS111_Data/ThuVienNhaDat.csv").drop(columns="Unnamed: 0")
df

,_id,link,post title,post date,location,price_per_m2,text,page,metadata.Date,metadata.Time,...,details.Nội thất,details.Mặt tiền,details.Đường vào,details.Chiều dài,details.Chiều rộng,details.Số tầng,details.Phí quản lý,details.Nằm ở tầng,details.Đơn vị thi công,details.Giá trên đơn vị tối thiểu/m²
0,6900b70eaeaa376b35f82fca,https://thuviennhadat.vn/ban-nha-rieng-thi-tra...,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,1,2025-10-28,19:29:02,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,6900b70eaeaa376b35f82fcb,https://thuviennhadat.vn/ban-can-ho-chung-cu-p...,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",1,2025-10-28,19:29:02,...,Cơ bản,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,6900b70eaeaa376b35f82fcc,https://thuviennhadat.vn/ban-shop-house-thi-tr...,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",1,2025-10-28,19:29:02,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,6900b70faeaa376b35f82fcd,https://thuviennhadat.vn/ban-can-ho-chung-cu-p...,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,1,2025-10-28,19:29:03,...,Đầy đủ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,6900b70faeaa376b35f82fce,https://thuviennhadat.vn/ban-dat-o-xa-tan-phu-...,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",1,2025-10-28,19:29:03,...,Không nội thất,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,6901668ea9a09ca97f38c617,https://thuviennhadat.vn/ban-nha-rieng-phuong-...,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",4215,2025-10-29,07:57:50,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
44359,6901668fa9a09ca97f38c618,https://thuviennhadat.vn/ban-nha-rieng-phuong-...,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",4215,2025-10-29,07:57:51,...,NaN,NaN,NaN,NaN,NaN,5 tầng,NaN,NaN,NaN,NaN
44360,6901668fa9a09ca97f38c619,https://thuviennhadat.vn/ban-dat-xa-thoi-tam-t...,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,4215,2025-10-29,07:57:51,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
44361,69016691a9a09ca97f38c61a,https://thuviennhadat.vn/ban-can-ho-chung-cu-x...,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",4216,2025-10-29,07:57:53,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# 3. Tiền xử lý dữ liệu

## a. Thống kê Missing Values

In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
_id,object,0,0.000000
link,object,0,0.000000
post title,object,21,0.047337
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,2366,5.333273
text,object,97,0.218651
page,int64,0,0.000000
metadata.Date,object,0,0.000000
metadata.Time,object,0,0.000000


Xử lý missing values

- Xoá cột không cần thiết + missing > 50%

- Xoá hàng có nội dung không thể điền + dự đoán (MNAR)

- Điền price_per_m2 bằng công thức  Mức giá / diện tích

- Pháp lý:

## b. Xoá cột không cần thiết

In [ ]:
df.drop(columns=[
    "_id",          # Mã lưu trữ từ CSDL MongoDB
    "link",         # Đường dẫn đến bài đăng chi tiết
    "page",         # Số trang trên web nơi bài viết được đăng
    "metadata.Time" # Giờ phút giây thu thập dữ liệu
], inplace=True)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Hướng nhà,...,details.Nội thất,details.Mặt tiền,details.Đường vào,details.Chiều dài,details.Chiều rộng,details.Số tầng,details.Phí quản lý,details.Nằm ở tầng,details.Đơn vị thi công,details.Giá trên đơn vị tối thiểu/m²
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,Nam,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,Tây Bắc,...,Cơ bản,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,...,Đầy đủ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,...,Không nội thất,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,Đông Nam,...,NaN,NaN,NaN,NaN,NaN,5 tầng,NaN,NaN,NaN,NaN
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,21,0.047337
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,2366,5.333273
text,object,97,0.218651
metadata.Date,object,0,0.000000
details.Mức giá,object,97,0.218651
details.Diện tích,object,97,0.218651
details.Pháp lý,object,10989,24.770642
details.Hướng nhà,object,34981,78.851746


## c. Xoá cột có tỉ lệ Missing Values > 50%

In [ ]:
df.drop(columns=[
    "details.Hướng nhà",
    "details.Hướng ban công",
    'details.Nội thất',
    'details.Mặt tiền',
    'details.Đường vào',
    'details.Chiều dài',
    'details.Chiều rộng',
    'details.Phí quản lý',
    'details.Nằm ở tầng',
    'details.Đơn vị thi công',
    'details.Giá trên đơn vị tối thiểu/m²',
    'details.Số tầng'
], inplace=True)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN


Thống kê Missing Values lại

In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values.head(20)

,Data_Type,Mising Values,Missing Percent
post title,object,21,0.047337
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,2366,5.333273
text,object,97,0.218651
metadata.Date,object,0,0.000000
details.Mức giá,object,97,0.218651
details.Diện tích,object,97,0.218651
details.Pháp lý,object,10989,24.770642
details.Số phòng ngủ,object,17895,40.337669


## d. Tạo thuộc tính ngày đăng gốc

**Giải thích:**

  + Thuộc tính *post date* : bài đăng chỉ cung cấp nội dung về thời gian đăng như "Đăng hôm nay", "Đăng hôm qua", "Đăng x ngày trước",... nên quá trình thống kê thời gian đăng bài gặp khó khăn
  + Thuộc tính *metadata.Date* : lưu ngày tháng năm thu thập dữ liệu

Như vậy thông qua hai thuộc tính này, nhóm đã tiến hành tạo ra thuộc tính ngày tháng năm đăng bài viết chính xác nhất

    --> **original_post_date**

In [ ]:
def compute_original_date(post_text, meta_date):
    """
    Xác định ngày đăng bài gốc (original post date) từ nội dung văn bản bài đăng
    và ngày metadata đi kèm.

    Hàm này trích xuất thông tin ngày đăng từ chuỗi `post_text` (thường là mô tả
    thời gian như "Đăng hôm nay", "Đăng 3 ngày trước", hoặc ngày cụ thể),
    sau đó quy đổi về một đối tượng datetime dựa trên `meta_date`.

    Các trường hợp được hỗ trợ:
    1. "Đăng hôm nay"
       → Trả về chính `meta_date`

    2. "Đăng hôm qua" / "Đăng hôm q"
       → Trả về `meta_date - 1 ngày`

    3. "Đăng x ngày trước" (x là số nguyên)
       → Trả về `meta_date - x ngày`

    4. Ngày dạng "dd/mm/yyyy"
       → Trả về datetime(y, m, d)

    5. Ngày dạng ISO "yyyy-mm-dd"
       → Trả về datetime(y, m, d)

    Thứ tự ưu tiên:
    - Các biểu thức thời gian tương đối (hôm nay, hôm qua, x ngày trước)
    - Sau đó mới đến ngày tuyệt đối (dd/mm/yyyy, yyyy-mm-dd)

    Parameters
    ----------
    post_text : str
        Chuỗi văn bản chứa thông tin ngày đăng bài
        (ví dụ: "Đăng 2 ngày trước", "Đăng hôm nay", "01/03/2023").

    meta_date : datetime hoặc str
        Ngày metadata của bài đăng (ví dụ: ngày crawl dữ liệu).
        Nếu là chuỗi, sẽ được chuyển sang datetime bằng `pd.to_datetime`.

    Returns
    -------
    datetime hoặc None
        - Trả về đối tượng `datetime` biểu diễn ngày đăng bài gốc nếu trích xuất thành công.
        - Trả về None nếu:
          + `meta_date` không hợp lệ
          + Không khớp bất kỳ mẫu ngày nào trong `post_text`.

    Notes
    -----
    - Hàm không xử lý múi giờ (timezone-naive datetime).
    - Văn bản `post_text` được chuyển về chữ thường (lowercase) trước khi xử lý.
    - Phù hợp để dùng trong tiền xử lý dữ liệu (feature engineering) cho bài toán NLP,
      crawling hoặc phân tích dữ liệu bất động sản / mạng xã hội.
    """
    # meta_date phải là datetime
    if isinstance(meta_date, str):
        meta_date = pd.to_datetime(meta_date, errors='coerce')
    if pd.isna(meta_date):
        return None

    text = str(post_text).strip().lower()
    # ---- "Đăng hôm nay" ----
    if "đăng hôm nay" in text:
        return meta_date

    # ---- "Đăng hôm qua" ----
    if "đăng hôm qua" in text or "đăng hôm q" in text:
        return meta_date - timedelta(days=1)

    # ---- "Đăng x ngày trước" ----
    match_days = re.search(r'đăng\s+(\d+)\s+ngày', text)
    if match_days:
        days = int(match_days.group(1))
        return meta_date - timedelta(days=days)

    # ---- "dd/mm/yyyy" ----
    match_ddmmyyyy = re.search(r'(\d{1,2})/(\d{1,2})/(\d{4})', text)
    if match_ddmmyyyy:
        d, m, y = match_ddmmyyyy.groups()
        return datetime(int(y), int(m), int(d))

    # ---- "yyyy-mm-dd" ----
    match_iso = re.search(r'(\d{4})-(\d{2})-(\d{2})', text)
    if match_iso:
        y, m, d = match_iso.groups()
        return datetime(int(y), int(m), int(d))

    return None

In [ ]:
df["original_post_date"] = df.apply(
    lambda row: compute_original_date(row["post date"], row["metadata.Date"]),
    axis=1
)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24
...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,21,0.047337
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,2366,5.333273
text,object,97,0.218651
metadata.Date,object,0,0.000000
details.Mức giá,object,97,0.218651
details.Diện tích,object,97,0.218651
details.Pháp lý,object,10989,24.770642
details.Số phòng ngủ,object,17895,40.337669


## e. Chuẩn hoá thuộc tính "price_per_m2"

In [ ]:
def convert_price_million(text):
    """
    Chuẩn hóa giá tiền về đơn vị triệu (million) từ chuỗi văn bản mô tả giá.

    Hàm này trích xuất giá trị số đầu tiên xuất hiện trong chuỗi `text`,
    sau đó quy đổi về đơn vị **triệu** dựa trên đơn vị được đề cập trong văn bản.

    Các trường hợp được hỗ trợ:
    1. Đơn vị "ngàn" / "ngan" (ví dụ: "850 ngàn/m2")
       → Giá trị được chia cho 1000 để chuyển sang triệu
       (850 ngàn = 0.85 triệu)

    2. Đơn vị "triệu" / "trieu" (ví dụ: "12.5 triệu/m2")
       → Giữ nguyên giá trị số

    3. Không xác định rõ đơn vị
       → Trả về giá trị số thô đã trích xuất (giả định đã ở đơn vị triệu)

    Parameters
    ----------
    text : str
        Chuỗi văn bản mô tả giá tiền
        (ví dụ: "12 triệu/m2", "850 ngàn/m2", "5.5 trieu").

    Returns
    -------
    float hoặc None
        - Giá trị giá tiền đã được chuẩn hóa về đơn vị triệu.
        - Trả về None nếu không tìm thấy giá trị số trong chuỗi.

    Notes
    -----
    - Hàm chỉ trích xuất **số đầu tiên** xuất hiện trong chuỗi.
    - Văn bản được chuyển sang chữ thường (lowercase) trước khi xử lý.
    - Hàm không phân biệt đơn vị theo diện tích (ví dụ: /m2), chỉ xử lý giá trị tiền.
    - Phù hợp dùng trong bước tiền xử lý dữ liệu (feature engineering)
      cho bài toán phân tích giá bất động sản.
    """
    t = str(text).lower()

    # Tách số
    match = re.search(r'\d+\.?\d*', t)
    if not match:
        return None

    num = float(match.group())

    # Nếu đơn vị là "ngàn/m2" → chia 1000 để chuyển sang triệu
    if "ngàn" in t or "ngan" in t:
        return num / 1000

    # Nếu đơn vị là "triệu/m2" → giữ nguyên
    if "triệu" in t or "trieu" in t:
        return num

    # Nếu không xác định rõ đơn vị → trả số thô
    return num

In [ ]:
df['price_per_m2_million'] = df['price_per_m2'].apply(convert_price_million)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26,17.04
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24,85.71
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24,170.62
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24,2.62
...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,1.18
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15,225.71
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,5.68
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,23.75


Sau khi thực hiện hàm **convert_price_million**, nhóm đã tạo được thuộc tính *price_per_m2_million* từ thuộc tính *price_per_m2*

In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,21,0.047337
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,2366,5.333273
text,object,97,0.218651
metadata.Date,object,0,0.000000
details.Mức giá,object,97,0.218651
details.Diện tích,object,97,0.218651
details.Pháp lý,object,10989,24.770642
details.Số phòng ngủ,object,17895,40.337669


## f. Chuẩn hoá thuộc tính "Mức giá"

In [ ]:
def convert_price(text):
    """
    Chuẩn hóa giá bất động sản về đơn vị **tỷ đồng** từ chuỗi văn bản mô tả giá.

    Hàm này trích xuất giá trị số đầu tiên xuất hiện trong chuỗi `text`,
    sau đó quy đổi về **đơn vị tỷ** dựa trên đơn vị tiền tệ được đề cập.

    Quy ước quy đổi:
    - "triệu"  → chia cho 1000 để chuyển sang tỷ
      (ví dụ: 500 triệu = 0.5 tỷ)
    - "tỷ" / "ty" → giữ nguyên giá trị
    - Không xác định được đơn vị → trả về None

    Parameters
    ----------
    text : str
        Chuỗi văn bản mô tả giá
        (ví dụ: "1.2 tỷ", "750 triệu", "3 ty").

    Returns
    -------
    float hoặc None
        - Giá trị giá đã được chuẩn hóa về **tỷ đồng**.
        - Trả về None nếu:
          + Không tìm thấy số trong chuỗi
          + Không xác định được đơn vị tiền tệ phù hợp.

    Notes
    -----
    - Hàm chỉ trích xuất **số đầu tiên** xuất hiện trong chuỗi.
    - Văn bản được xử lý không phân biệt chữ hoa / chữ thường.
    - Hàm không xử lý các trường hợp phức tạp như:
        + "1 tỷ 200 triệu"
        + "khoảng 2–3 tỷ"
        + "giá thỏa thuận"
    - Phù hợp dùng trong bước tiền xử lý dữ liệu (feature engineering)
      cho các bài toán phân tích giá bất động sản.
    """

    # Tìm số
    match = re.search(r'\d+\.?\d*', str(text))
    if not match:
        return None   # không có số → trả về None

    num = float(match.group())

    # Quy đổi
    t = str(text).lower()

    if "triệu" in t:
        return num / 1000

    if "tỷ" in t or "ty" in t:
        return num

    return None

In [ ]:
# Tạo cột mới: giá trị quy đổi về Tỷ
df["price"] = df["details.Mức giá"].apply(convert_price)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26,17.04,6.90
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24,85.71,9.00
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24,170.62,36.00
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24,2.62,0.63
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,1.18,0.40
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15,225.71,7.90
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,5.68,6.70
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,23.75,0.95


Sau khi thực hiện hàm **convert_price**, nhóm đã tạo được thuộc tính *price* từ thuộc tính *details.Mức giá* (đơn vị: tỷ VNĐ)

Ngoài ra, sau khi tạo thuộc tính **price**, nhóm còn phát hiện có một số bài đăng không có mức giá chi tiết, thay vào đó là Mức giá "thoả thuận".

Vì vậy, nhóm quuyết định loại bỏ các dòng dữ liệu chứa Mức giá "thoả thuận"

In [ ]:
df["details.Mức giá"].value_counts()

,count
details.Mức giá,
Thỏa thuận,2269
6.50 tỷ,465
6 tỷ,428
7.50 tỷ,413
7 tỷ,381
...,...
285.60 triệu,1
9.91 tỷ,1
442.40 triệu,1


In [ ]:
df = df[~df["details.Mức giá"].str.contains("thỏa thuận", case=False, na=False)]
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26,17.04,6.90
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24,85.71,9.00
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24,170.62,36.00
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24,2.62,0.63
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,1.18,0.40
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15,225.71,7.90
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,5.68,6.70
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,23.75,0.95


Cuối cùng, loại bỏ các dòng có mức giá không rõ ràng, chỉ chứa "VND" và số tiền không xác định rõ ràng.

In [ ]:
# Lọc ra các dòng có chứa "VND" trong thuộc tính "details.Mức giá"
df_vnd = df[df["details.Mức giá"].str.contains("VND", case=False, na=False)]
df_vnd

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price
182,"Bán nhà Hoàng Hoa Thám, hẻm xe hơi 1 trục, 44m...",Đăng hôm nay,"Phú Nhuận, Hồ Chí Minh",~0 ngàn/m2,"Bán nhà Hoàng Hoa Thám, hẻm xe hơi 1 trục, 44m...",2025-10-28,645 VND,44 m²,Sổ đỏ / Sổ hồng,4 phòng,4 phòng,2025-10-28,0.0,NaN
12165,"Bán căn hộ Melia Nha Trang, nghỉ dưỡng đẳng cấ...",Đăng 15/10/2025,"Nha Trang, Khánh Hòa",~0 ngàn/m2,Căn hộ Melias Nha Trang - căn hộ đẳng cấp 5sao...,2025-10-28,"9,000 VND",90 m²,Sổ đỏ / Sổ hồng,3 phòng,3 phòng,2025-10-15,0.0,NaN
27853,"Bán nhà riêng 113,5m2 tại Thịnh Quang, Đống Đa...",Đăng 01/04/2025,"Đống Đa, Hà Nội",~0 ngàn/m2,Thông tin mô tảNhà riêng tọa lạc tại Thịnh Qua...,2025-10-28,"34,617 VND",113.5 m²,Sổ đỏ / Sổ hồng,5 phòng,5 phòng,2025-04-01,0.0,NaN
33704,"Cần bán đất 20.400m² mặt tiền đường DT 716, Hồ...",Đăng 25/03/2025,"Bắc Bình, Bình Thuận",~0 ngàn/m2,"Tin bán đất mặt tiền đường DT 716, khu vực Hồn...",2025-10-28,"779,000 VND",20400 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-25,0.0,NaN
34059,Bán đất 10.000 m2 mặt tiền đường Nguyễn Bách x...,Đăng 25/03/2025,"Định Quán, Đồng Nai",~0 ngàn/m2,Đây là tin bán đất 10.000 m2 thuộc mặt tiền đư...,2025-10-28,"750,000 VND",10000 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-25,0.0,NaN
40932,Bán căn 2n siêu đẹp rẻ nhất dự án giá 4.099 tỷ,Đăng 19/03/2025,"Nam Từ Liêm, Hà Nội",~0 ngàn/m2,"Tôi chuyên mua bán, đầu tư, vận hành cho thuê ...",2025-10-28,"4,099 VND",64 m²,Sổ đỏ / Sổ hồng,2 phòng,2 phòng,2025-03-19,0.0,NaN


In [ ]:
# Chỉ chọn các dòng không chứa "VND" trong thuộc tính "details.Mức giá" --> Cũng là đang làm sạch cho thuộc tính "price"
df = df[~df["details.Mức giá"].str.contains("VND", case=False, na=False)]
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26,17.04,6.90
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24,85.71,9.00
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24,170.62,36.00
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24,2.62,0.63
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,1.18,0.40
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15,225.71,7.90
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,5.68,6.70
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,23.75,0.95


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,21,0.049895
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,97,0.230469
text,object,97,0.230469
metadata.Date,object,0,0.000000
details.Mức giá,object,97,0.230469
details.Diện tích,object,97,0.230469
details.Pháp lý,object,10296,24.463030
details.Số phòng ngủ,object,16766,39.835583


## g. Chuẩn hoá thuộc tính "Diện tích"

In [ ]:
def convert_area(text):
    """
    Trích xuất và chuẩn hóa diện tích từ chuỗi văn bản.

    Hàm này tìm **giá trị số đầu tiên** xuất hiện trong chuỗi `text`
    và trả về dưới dạng số thực (float), thường dùng để biểu diễn
    diện tích (m²).

    Parameters
    ----------
    text : str
        Chuỗi văn bản mô tả diện tích
        (ví dụ: "82m2", "diện tích 45.5 m²", "khoảng 100 mét vuông").

    Returns
    -------
    float hoặc None
        - Giá trị diện tích đã trích xuất (float).
        - Trả về None nếu không tìm thấy giá trị số trong chuỗi.

    Notes
    -----
    - Hàm không kiểm tra đơn vị diện tích (m2, m², mét vuông...).
    - Chỉ lấy **số đầu tiên** xuất hiện trong chuỗi.
    - Không xử lý các trường hợp phức tạp như:
        + "80–100m2"
        + "2 x 5 = 10m2"
    - Phù hợp dùng trong bước tiền xử lý dữ liệu (feature engineering)
      cho các bài toán phân tích bất động sản.
    """
    match = re.search(r'\d+\.?\d*', str(text))
    return float(match.group()) if match else None

In [ ]:
df['area'] = df['details.Diện tích'].apply(convert_area)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price,area
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26,17.04,6.90,405.0
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24,85.71,9.00,105.0
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24,170.62,36.00,211.0
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24,2.62,0.63,240.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,1.18,0.40,340.0
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15,225.71,7.90,35.0
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,5.68,6.70,1179.0
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,23.75,0.95,40.0


## h. Tạo thuộc tính mới từ "location"

Từ thuộc tính "location", tách thành 2 thuộc tính mới là **ward/district** (chứa tên phường/quận) và **city/province** (chứa tên tỉnh/thành)

In [ ]:
df[['ward/district', 'city/province']] = df['location'].str.split(',', n=1, expand=True)

# Xóa khoảng trắng dư thừa
df['ward/district'] = df['ward/district'].str.strip()
df['city/province'] = df['city/province'].str.strip()
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price,area,ward/district,city/province
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26,17.04,6.90,405.0,Bến Lức,Long An
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24,85.71,9.00,105.0,Long Thành,Đồng Nai
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24,170.62,36.00,211.0,Thủ Đức,Hồ Chí Minh
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24,2.62,0.63,240.0,Củ Chi,Hồ Chí Minh
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,1.18,0.40,340.0,Bến Cát,Bình Dương
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15,225.71,7.90,35.0,Tây Hồ,Hà Nội
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,5.68,6.70,1179.0,Hóc Môn,Hồ Chí Minh
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,23.75,0.95,40.0,Bến Lức,Long An


In [ ]:
df["city/province"].value_counts()

,count
city/province,
Hồ Chí Minh,16512
Hà Nội,13915
Bình Dương,1912
Đà Nẵng,1778
Khánh Hòa,1470
Long An,1071
Đồng Nai,1002
Lâm Đồng,398
Cần Thơ,372


## i. Xử lý Missing Values cột "post title"

- **Hướng xử lý:** Xoá tất cả các hàng có chứa Missing Values (tổng cộng 21 hàng)

- **Lý do:** "post title" là tên tiêu đề của các bài đăng, không thể xử lý bằng cách điền khuyết, hay mô hình dự đoán (Missing Not At Random - MNAR)

In [ ]:
df.dropna(
    axis=0,
    how='any',
    subset=[
        "post title"
    ],
    inplace=True)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price,area,ward/district,city/province
0,"Bán tòa nhà Phước Toàn, Long An, gần chợ Gò Đe...",Đăng 2 ngày trước,"Bến Lức, Long An",~17.04 triệu/m2,Chính chủ bán tòa nhà phòng trọ tại đường Phướ...,2025-10-28,6.90 tỷ,405 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-26,17.04,6.90,405.0,Bến Lức,Long An
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương
2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",Đăng 4 ngày trước,"Long Thành, Đồng Nai",~85.71 triệu/m2,"Bán Shophouse ID Junction, Long Thành, Đồng Na...",2025-10-28,9 tỷ,105 m²,Hợp đồng mua bán,NaN,NaN,2025-10-24,85.71,9.00,105.0,Long Thành,Đồng Nai
3,"Bán căn hộ Penthouse Sadora Thủ Thiêm, view La...",Đăng 4 ngày trước,"Thủ Đức, Hồ Chí Minh",~170.62 triệu/m2,Sở hữu không gian sống đỉnh cao tại Penthouse ...,2025-10-28,36 tỷ,211 m²,NaN,NaN,NaN,2025-10-24,170.62,36.00,211.0,Thủ Đức,Hồ Chí Minh
4,"CHÍNH CHỦ CẦN BÁN ĐẤT NGAY BỆNH VIỆN XUYÊN Á, ...",Đăng 4 ngày trước,"Củ Chi, Hồ Chí Minh",~2.62 triệu/m2,"Cam kết tin đăng không ảo, không bị quy hoạchC...",2025-10-28,630 triệu,240 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-10-24,2.62,0.63,240.0,Củ Chi,Hồ Chí Minh
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44358,"Cần bán lô đất sổ sẵn cách chợ chánh lưu 800m,...",Đăng 15/03/2025,"Bến Cát, Bình Dương",~1.18 triệu/m2,"ĐẤT NỀN KCN VSIP 2 BÌNH DƯƠNG, cách chợ Chánh ...",2025-10-29,400 triệu,340 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,1.18,0.40,340.0,Bến Cát,Bình Dương
44359,"Bán nhà riêng 35m2 x 5 tầng Hoàng Hoa Thám, Tâ...",Đăng 15/03/2025,"Tây Hồ, Hà Nội",~225.71 triệu/m2,"Tin bán nhà riêng tại phố Hoàng Hoa Thám, phườ...",2025-10-29,7.90 tỷ,35 m²,NaN,3 phòng,NaN,2025-03-15,225.71,7.90,35.0,Tây Hồ,Hà Nội
44360,Bán đất mặt tiền đường nhựa trung đông 7 - thớ...,Đăng 15/03/2025,"Hóc Môn, Hồ Chí Minh",~5.68 triệu/m2,Bán đất mặt tiền đường nhựa Trung Đông 7 Đối D...,2025-10-29,6.70 tỷ,1179 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,5.68,6.70,1179.0,Hóc Môn,Hồ Chí Minh
44361,Bán căn hộ chung cư 2PN xã Mỹ Yên huyện Bến Lứ...,Đăng 15/03/2025,"Bến Lức, Long An",~23.75 triệu/m2,"Bán căn hộ chung cư tại xã Mỹ Yên, huyện Bến L...",2025-10-29,950 triệu,40 m²,Sổ đỏ / Sổ hồng,NaN,NaN,2025-03-15,23.75,0.95,40.0,Bến Lức,Long An


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,0,0.000000
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,76,0.180664
text,object,76,0.180664
metadata.Date,object,0,0.000000
details.Mức giá,object,76,0.180664
details.Diện tích,object,76,0.180664
details.Pháp lý,object,10275,24.425322
details.Số phòng ngủ,object,16745,39.805548


## j. Xử lý Missing Values cột "Số phòng ngủ" và "Số phòng WC"

- **Hướng xử lý:** Xoá tất cả các hàng có chứa Missing Values ở cả hai thuộc tính "details.Số phòng ngủ" và "details.Số phòng WC"

- **Lý do:** Cả hai thuộc tính "details.Số phòng ngủ" và "details.Số phòng WC" đều là thông tin cần tính chính xác cao, giá trị thiếu rất nhiều (16745 và 18525) nên  không thể xử lý bằng cách điền khuyết, hay mô hình dự đoán.

In [ ]:
df.dropna(
    axis=0,
    how='any',
    subset=[
        'details.Số phòng ngủ', 'details.Số phòng WC',
    ],
    inplace=True)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price,area,ward/district,city/province
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương
6,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",Đăng hôm qua,"Quận 10, Hồ Chí Minh",~50 triệu/m2,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",2025-10-28,2.65 tỷ,53 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-27,50.00,2.65,53.0,Quận 10,Hồ Chí Minh
9,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",Đăng 3 ngày trước,"Bình Tân, Hồ Chí Minh",~92.71 triệu/m2,"Chào bạn, mình đang cần bán căn nhà riêng tọa ...",2025-10-28,4.45 tỷ,48 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,92.71,4.45,48.0,Bình Tân,Hồ Chí Minh
10,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",Đăng 3 ngày trước,"Thủ Đức, Hồ Chí Minh",~59.6 triệu/m2,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",2025-10-28,5.90 tỷ,99 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,59.60,5.90,99.0,Thủ Đức,Hồ Chí Minh
13,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",Đăng 4 ngày trước,"Bình Tân, Hồ Chí Minh",~110.94 triệu/m2,​🏡 NHÀ PHỐ CAO CẤP - CHIẾN LƯỢC - 5 TẦNG - BÌN...,2025-10-28,7.10 tỷ,64 m²,Sổ đỏ / Sổ hồng,5 phòng,5 phòng,2025-10-24,110.94,7.10,64.0,Bình Tân,Hồ Chí Minh
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44350,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",Đăng 15/03/2025,"Thủ Đức, Hồ Chí Minh",~97.5 triệu/m2,"Cần bán nhà trệt 2 lầu ngay đường số 8 , Phườn...",2025-10-29,5.85 tỷ,60 m²,NaN,4 phòng,4 phòng,2025-03-15,97.50,5.85,60.0,Thủ Đức,Hồ Chí Minh
44351,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,Đăng 15/03/2025,"Tân Phú, Hồ Chí Minh",~76.43 triệu/m2,"Bán nhà đường Tây Thạnh, phường Tây Thạnh, Quậ...",2025-10-29,4.28 tỷ,56 m²,NaN,2 phòng,2 phòng,2025-03-15,76.43,4.28,56.0,Tân Phú,Hồ Chí Minh
44352,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,Đăng 15/03/2025,"Tây Hồ, Hà Nội",~238.33 triệu/m2,"Sang nhượng nhà riêng tại Lạc Long Quân, Phườn...",2025-10-29,8.58 tỷ,36 m²,Sổ đỏ / Sổ hồng,3 phòng,3 phòng,2025-03-15,238.33,8.58,36.0,Tây Hồ,Hà Nội
44353,"Bán nhà riêng 62,5m² hẻm đường Âu Cơ, phường P...",Đăng 15/03/2025,"Nha Trang, Khánh Hòa",~90.32 triệu/m2,Đây là tin bán nhà riêng với thiết kế 3 tầng t...,2025-10-29,5.60 tỷ,62 m²,NaN,2 phòng,3 phòng,2025-03-15,90.32,5.60,62.0,Nha Trang,Khánh Hòa


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,0,0.000000
post date,object,0,0.000000
location,object,0,0.000000
price_per_m2,object,0,0.000000
text,object,0,0.000000
metadata.Date,object,0,0.000000
details.Mức giá,object,0,0.000000
details.Diện tích,object,0,0.000000
details.Pháp lý,object,3839,16.577425
details.Số phòng ngủ,object,0,0.000000


## k. Xử lý Missing Values cột "Pháp lý"

- **Hướng xử lý:**

    + Bước 1: Tạo thuộc tính " " - đánh dấu trong phần mô tả nhà đất có đề cập đến vấn đề Pháp lý hay không (0 là không có, 1 là có)

    + Bước 2: Tìm trong text có chứa các cụm từ mà cột Pháp lý có như "Sổ đỏ", "Sổ hồng", "Hợp đồng mua bán", "Đang chờ sổ", "Giấy viết tay"

    + Bước 3: Nếu mẫu nào không có thì không điền, nếu mẫu nào có thì điền theo các nguyên tắc sau:

        - "Sổ hồng" --> "Sổ đỏ / Sổ hồng"

        - "Sổ đỏ" --> "Sổ đỏ / Sổ hồng"

        - "Hợp đồng mua bán" --> "Hợp đồng mua bán"

        - "Đang chờ sổ" --> "Đang chờ sổ"

        - "Giấy viết tay" --> "Giấy viết tay"
    
    + Bước 4: Xoá tất cả mẫu dữ liệu sau khi đã thực hiện lọc ra nhưng vẫn NULL

Liệt kê các giá trị duy nhất

In [ ]:
df["details.Pháp lý"].value_counts()

,count
details.Pháp lý,
Sổ đỏ / Sổ hồng,17609
Hợp đồng mua bán,1647
Đang chờ sổ,50
Giấy viết tay,13


### Bước 1.

In [ ]:
df["is_legal"] = df["text"].str.lower().str.contains("pháp lý", na=False).astype(int)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price,area,ward/district,city/province,is_legal
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương,1
6,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",Đăng hôm qua,"Quận 10, Hồ Chí Minh",~50 triệu/m2,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",2025-10-28,2.65 tỷ,53 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-27,50.00,2.65,53.0,Quận 10,Hồ Chí Minh,1
9,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",Đăng 3 ngày trước,"Bình Tân, Hồ Chí Minh",~92.71 triệu/m2,"Chào bạn, mình đang cần bán căn nhà riêng tọa ...",2025-10-28,4.45 tỷ,48 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,92.71,4.45,48.0,Bình Tân,Hồ Chí Minh,1
10,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",Đăng 3 ngày trước,"Thủ Đức, Hồ Chí Minh",~59.6 triệu/m2,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",2025-10-28,5.90 tỷ,99 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,59.60,5.90,99.0,Thủ Đức,Hồ Chí Minh,0
13,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",Đăng 4 ngày trước,"Bình Tân, Hồ Chí Minh",~110.94 triệu/m2,​🏡 NHÀ PHỐ CAO CẤP - CHIẾN LƯỢC - 5 TẦNG - BÌN...,2025-10-28,7.10 tỷ,64 m²,Sổ đỏ / Sổ hồng,5 phòng,5 phòng,2025-10-24,110.94,7.10,64.0,Bình Tân,Hồ Chí Minh,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44350,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",Đăng 15/03/2025,"Thủ Đức, Hồ Chí Minh",~97.5 triệu/m2,"Cần bán nhà trệt 2 lầu ngay đường số 8 , Phườn...",2025-10-29,5.85 tỷ,60 m²,NaN,4 phòng,4 phòng,2025-03-15,97.50,5.85,60.0,Thủ Đức,Hồ Chí Minh,1
44351,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,Đăng 15/03/2025,"Tân Phú, Hồ Chí Minh",~76.43 triệu/m2,"Bán nhà đường Tây Thạnh, phường Tây Thạnh, Quậ...",2025-10-29,4.28 tỷ,56 m²,NaN,2 phòng,2 phòng,2025-03-15,76.43,4.28,56.0,Tân Phú,Hồ Chí Minh,1
44352,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,Đăng 15/03/2025,"Tây Hồ, Hà Nội",~238.33 triệu/m2,"Sang nhượng nhà riêng tại Lạc Long Quân, Phườn...",2025-10-29,8.58 tỷ,36 m²,Sổ đỏ / Sổ hồng,3 phòng,3 phòng,2025-03-15,238.33,8.58,36.0,Tây Hồ,Hà Nội,0
44353,"Bán nhà riêng 62,5m² hẻm đường Âu Cơ, phường P...",Đăng 15/03/2025,"Nha Trang, Khánh Hòa",~90.32 triệu/m2,Đây là tin bán nhà riêng với thiết kế 3 tầng t...,2025-10-29,5.60 tỷ,62 m²,NaN,2 phòng,3 phòng,2025-03-15,90.32,5.60,62.0,Nha Trang,Khánh Hòa,1


### Bước 2.

In [ ]:
def classify_legal(text):
    """
    Phân loại tình trạng pháp lý bất động sản từ mô tả văn bản.

    Hàm này dựa trên việc dò tìm các từ khóa pháp lý phổ biến trong chuỗi `text`
    để gán nhãn tình trạng pháp lý của bất động sản theo các nhóm chuẩn hóa.

    Các từ khóa được nhận diện:
    - "sổ hồng", "sổ đỏ"
    - "hợp đồng mua bán"
    - "giấy viết tay"
    - "đang chờ sổ"

    Thứ tự ưu tiên phân loại (từ cao xuống thấp):
    1. Có "sổ hồng" hoặc "sổ đỏ"
       → "Sổ đỏ / Sổ hồng"

    2. Không có sổ, nhưng có "hợp đồng mua bán"
       → "Hợp đồng mua bán"

    3. Không có sổ, không có HĐMB, nhưng có "giấy viết tay"
       → "Giấy viết tay"

    4. Không có các trường hợp trên, nhưng có "đang chờ sổ"
       → "Đang chờ sổ"

    5. Không xác định được
       → NaN

    Parameters
    ----------
    text : str hoặc None
        Chuỗi mô tả pháp lý của bất động sản
        (ví dụ: "Nhà có sổ hồng riêng", "Hợp đồng mua bán, đang chờ sổ").

    Returns
    -------
    str hoặc numpy.nan
        Nhãn pháp lý đã được chuẩn hóa, hoặc NaN nếu không xác định được.

    Notes
    -----
    - Văn bản được chuyển về chữ thường (lowercase) trước khi xử lý.
    - Hàm chỉ dựa trên từ khóa, không phân tích ngữ nghĩa sâu.
    - Trường hợp văn bản chứa nhiều từ khóa,
      kết quả phụ thuộc vào **thứ tự ưu tiên** đã định nghĩa.
    - Phù hợp dùng trong bước tiền xử lý dữ liệu (feature engineering)
      cho các bài toán phân tích bất động sản.
    """

    if text is None:
        return np.nan

    t = text.lower()

    has_so_hong = "sổ hồng" in t
    has_so_do   = "sổ đỏ" in t
    has_hdmua   = "hợp đồng mua bán" in t
    has_giayvt  = "giấy viết tay" in t
    has_dangcho = "đang chờ sổ" in t

    # Ưu tiên: nếu có sổ hồng / sổ đỏ → gán Sổ đỏ / Sổ hồng
    if has_so_hong or has_so_do:
        return "Sổ đỏ / Sổ hồng"

    # Nếu không có sổ mà có hợp đồng mua bán
    if has_hdmua:
        return "Hợp đồng mua bán"

    # Nếu không có sổ, không có HĐMB mà có giấy viết tay
    if has_giayvt:
        return "Giấy viết tay"

    if has_dangcho:
        return "Đang chờ sổ"

    # Không có gì → giữ NaN
    return np.nan

### Bước 3.

In [ ]:
# Tạo cột mới "details.Pháp lý"
mask = df["details.Pháp lý"].isna()

df.loc[mask, "details.Pháp lý"] = df.loc[mask, "text"].apply(classify_legal)

In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,0,0.000
post date,object,0,0.000
location,object,0,0.000
price_per_m2,object,0,0.000
text,object,0,0.000
metadata.Date,object,0,0.000
details.Mức giá,object,0,0.000
details.Diện tích,object,0,0.000
details.Pháp lý,object,2307,9.962
details.Số phòng ngủ,object,0,0.000


**Nhận xét** : Sau khi thực hiện hàm **classify_legal** để điền khuyết, giá trị thiếu của thuộc tính "details.Pháp lý" đã giảm từ 3839 xuống còn 2307.

### Bước 4.

In [ ]:
# Xoá tất cả dòng còn khuyết còn lại
df.dropna(
    axis=0,
    how='any',
    subset=[
        'details.Pháp lý'
    ],
    inplace=True)
df

,post title,post date,location,price_per_m2,text,metadata.Date,details.Mức giá,details.Diện tích,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price,area,ward/district,city/province,is_legal
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Đăng 4 ngày trước,"Dĩ An, Bình Dương",~37.8 triệu/m2,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-28,3.10 tỷ,82 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương,1
6,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",Đăng hôm qua,"Quận 10, Hồ Chí Minh",~50 triệu/m2,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",2025-10-28,2.65 tỷ,53 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-27,50.00,2.65,53.0,Quận 10,Hồ Chí Minh,1
9,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",Đăng 3 ngày trước,"Bình Tân, Hồ Chí Minh",~92.71 triệu/m2,"Chào bạn, mình đang cần bán căn nhà riêng tọa ...",2025-10-28,4.45 tỷ,48 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,92.71,4.45,48.0,Bình Tân,Hồ Chí Minh,1
10,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",Đăng 3 ngày trước,"Thủ Đức, Hồ Chí Minh",~59.6 triệu/m2,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",2025-10-28,5.90 tỷ,99 m²,Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,59.60,5.90,99.0,Thủ Đức,Hồ Chí Minh,0
13,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",Đăng 4 ngày trước,"Bình Tân, Hồ Chí Minh",~110.94 triệu/m2,​🏡 NHÀ PHỐ CAO CẤP - CHIẾN LƯỢC - 5 TẦNG - BÌN...,2025-10-28,7.10 tỷ,64 m²,Sổ đỏ / Sổ hồng,5 phòng,5 phòng,2025-10-24,110.94,7.10,64.0,Bình Tân,Hồ Chí Minh,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44348,"Bán nhà riêng 108 m2 đường Phan Văn Hớn, Tân T...",Đăng 15/03/2025,"Quận 12, Hồ Chí Minh",~14.35 triệu/m2,Đây là tin bán nhà riêng tại phường Tân Thới N...,2025-10-29,1.55 tỷ,108 m²,Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-03-15,14.35,1.55,108.0,Quận 12,Hồ Chí Minh,1
44350,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",Đăng 15/03/2025,"Thủ Đức, Hồ Chí Minh",~97.5 triệu/m2,"Cần bán nhà trệt 2 lầu ngay đường số 8 , Phườn...",2025-10-29,5.85 tỷ,60 m²,Sổ đỏ / Sổ hồng,4 phòng,4 phòng,2025-03-15,97.50,5.85,60.0,Thủ Đức,Hồ Chí Minh,1
44351,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,Đăng 15/03/2025,"Tân Phú, Hồ Chí Minh",~76.43 triệu/m2,"Bán nhà đường Tây Thạnh, phường Tây Thạnh, Quậ...",2025-10-29,4.28 tỷ,56 m²,Sổ đỏ / Sổ hồng,2 phòng,2 phòng,2025-03-15,76.43,4.28,56.0,Tân Phú,Hồ Chí Minh,1
44352,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,Đăng 15/03/2025,"Tây Hồ, Hà Nội",~238.33 triệu/m2,"Sang nhượng nhà riêng tại Lạc Long Quân, Phườn...",2025-10-29,8.58 tỷ,36 m²,Sổ đỏ / Sổ hồng,3 phòng,3 phòng,2025-03-15,238.33,8.58,36.0,Tây Hồ,Hà Nội,0


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
post title,object,0,0.0
post date,object,0,0.0
location,object,0,0.0
price_per_m2,object,0,0.0
text,object,0,0.0
metadata.Date,object,0,0.0
details.Mức giá,object,0,0.0
details.Diện tích,object,0,0.0
details.Pháp lý,object,0,0.0
details.Số phòng ngủ,object,0,0.0


## l. Đổi tên và xoá các thuộc tính không cần thiết

In [ ]:
df.columns

Index(['post title', 'post date', 'location', 'price_per_m2', 'text',
       'metadata.Date', 'details.Mức giá', 'details.Diện tích',
       'details.Pháp lý', 'details.Số phòng ngủ', 'details.Số phòng WC',
       'original_post_date', 'price_per_m2_million', 'price', 'area',
       'ward/district', 'city/province', 'is_legal'],
      dtype='object')

In [ ]:
df = df.drop(columns=[
    'post date', 'location', 'price_per_m2', 'text',
    'metadata.Date', 'details.Mức giá', 'details.Diện tích'
])
df

,post title,details.Pháp lý,details.Số phòng ngủ,details.Số phòng WC,original_post_date,price_per_m2_million,price,area,ward/district,city/province,is_legal
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương,1
6,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-27,50.00,2.65,53.0,Quận 10,Hồ Chí Minh,1
9,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,92.71,4.45,48.0,Bình Tân,Hồ Chí Minh,1
10,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,59.60,5.90,99.0,Thủ Đức,Hồ Chí Minh,0
13,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",Sổ đỏ / Sổ hồng,5 phòng,5 phòng,2025-10-24,110.94,7.10,64.0,Bình Tân,Hồ Chí Minh,0
...,...,...,...,...,...,...,...,...,...,...,...
44348,"Bán nhà riêng 108 m2 đường Phan Văn Hớn, Tân T...",Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-03-15,14.35,1.55,108.0,Quận 12,Hồ Chí Minh,1
44350,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",Sổ đỏ / Sổ hồng,4 phòng,4 phòng,2025-03-15,97.50,5.85,60.0,Thủ Đức,Hồ Chí Minh,1
44351,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,Sổ đỏ / Sổ hồng,2 phòng,2 phòng,2025-03-15,76.43,4.28,56.0,Tân Phú,Hồ Chí Minh,1
44352,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,Sổ đỏ / Sổ hồng,3 phòng,3 phòng,2025-03-15,238.33,8.58,36.0,Tây Hồ,Hà Nội,0


In [ ]:
df.columns

Index(['post title', 'details.Pháp lý', 'details.Số phòng ngủ',
       'details.Số phòng WC', 'original_post_date', 'price_per_m2_million',
       'price', 'area', 'ward/district', 'city/province', 'is_legal'],
      dtype='object')

In [ ]:
df = df.rename(columns={
    "post title" : "title",
    'original_post_date': 'original_post_date',
    'ward/district' : 'ward/district',
    'city/province' : 'city/province',
    'details.Số phòng WC': 'total_restroom',
    'details.Số phòng ngủ': 'total_bedroom',
    'details.Pháp lý': 'legal_doc',
    'price': 'price',
    'price_per_m2_million': 'price_per_m2_million',
})
df

,title,legal_doc,total_bedroom,total_restroom,original_post_date,price_per_m2_million,price,area,ward/district,city/province,is_legal
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương,1
6,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-27,50.00,2.65,53.0,Quận 10,Hồ Chí Minh,1
9,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,92.71,4.45,48.0,Bình Tân,Hồ Chí Minh,1
10,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,59.60,5.90,99.0,Thủ Đức,Hồ Chí Minh,0
13,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",Sổ đỏ / Sổ hồng,5 phòng,5 phòng,2025-10-24,110.94,7.10,64.0,Bình Tân,Hồ Chí Minh,0
...,...,...,...,...,...,...,...,...,...,...,...
44348,"Bán nhà riêng 108 m2 đường Phan Văn Hớn, Tân T...",Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-03-15,14.35,1.55,108.0,Quận 12,Hồ Chí Minh,1
44350,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",Sổ đỏ / Sổ hồng,4 phòng,4 phòng,2025-03-15,97.50,5.85,60.0,Thủ Đức,Hồ Chí Minh,1
44351,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,Sổ đỏ / Sổ hồng,2 phòng,2 phòng,2025-03-15,76.43,4.28,56.0,Tân Phú,Hồ Chí Minh,1
44352,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,Sổ đỏ / Sổ hồng,3 phòng,3 phòng,2025-03-15,238.33,8.58,36.0,Tây Hồ,Hà Nội,0


## m. Điền Missing Value cho cột "price_per_m2_million"

Mặc dù giá trị thiếu của thuộc tính **price_per_m2_million** đã vô tình được xoá đi sau khi xử lý giá trị thiếu cho "details.Số phòng ngủ" và "details.Số phòng WC", nhóm vẫn tiến hành đưa ra phương án xử lý giá trị thiếu cho thuộc tính này để xử lý cho trường hợp có dữ liệu mới được thêm vào

Điền giá trị thiếu cho cột "price_per_m2_million" bằng công thức:
**price_per_m2_million = price / area**

In [ ]:
mask = df["price_per_m2_million"].isna()

df.loc[mask, "price_per_m2_million"] = (
    df.loc[mask, "price"] / df.loc[mask, "area"]
)
df

,title,legal_doc,total_bedroom,total_restroom,original_post_date,price_per_m2_million,price,area,ward/district,city/province,is_legal
1,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương,1
6,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-27,50.00,2.65,53.0,Quận 10,Hồ Chí Minh,1
9,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,92.71,4.45,48.0,Bình Tân,Hồ Chí Minh,1
10,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",Sổ đỏ / Sổ hồng,2 phòng,1 phòng,2025-10-25,59.60,5.90,99.0,Thủ Đức,Hồ Chí Minh,0
13,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",Sổ đỏ / Sổ hồng,5 phòng,5 phòng,2025-10-24,110.94,7.10,64.0,Bình Tân,Hồ Chí Minh,0
...,...,...,...,...,...,...,...,...,...,...,...
44348,"Bán nhà riêng 108 m2 đường Phan Văn Hớn, Tân T...",Sổ đỏ / Sổ hồng,3 phòng,2 phòng,2025-03-15,14.35,1.55,108.0,Quận 12,Hồ Chí Minh,1
44350,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",Sổ đỏ / Sổ hồng,4 phòng,4 phòng,2025-03-15,97.50,5.85,60.0,Thủ Đức,Hồ Chí Minh,1
44351,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,Sổ đỏ / Sổ hồng,2 phòng,2 phòng,2025-03-15,76.43,4.28,56.0,Tân Phú,Hồ Chí Minh,1
44352,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,Sổ đỏ / Sổ hồng,3 phòng,3 phòng,2025-03-15,238.33,8.58,36.0,Tây Hồ,Hà Nội,0


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
title,object,0,0.0
legal_doc,object,0,0.0
total_bedroom,object,0,0.0
total_restroom,object,0,0.0
original_post_date,datetime64[ns],0,0.0
price_per_m2_million,float64,0,0.0
price,float64,0,0.0
area,float64,0,0.0
ward/district,object,0,0.0
city/province,object,0,0.0


## n. Chuẩn hoá thuộc tính "total_bedroom" và "total_restroom"

In [ ]:
def extract_number(text):
    """
    Trích xuất số nguyên đầu tiên xuất hiện trong chuỗi văn bản.

    Hàm này tìm và lấy **chuỗi chữ số liên tiếp đầu tiên** trong `text`,
    sau đó chuyển sang kiểu số nguyên (`int`).
    Thường được dùng để chuẩn hóa các cột dạng mô tả
    như số phòng ngủ, số nhà vệ sinh.

    Parameters
    ----------
    text : str
        Chuỗi văn bản chứa số
        (ví dụ: "3 phòng ngủ", "WC: 2", "1 toilet").

    Returns
    -------
    int hoặc None
        - Giá trị số nguyên đã trích xuất.
        - Trả về None nếu không tìm thấy số trong chuỗi.

    Notes
    -----
    - Chỉ trích xuất **số đầu tiên** xuất hiện trong chuỗi.
    - Không xử lý số thập phân (ví dụ: "1.5").
    - Không xử lý các trường hợp phức tạp như:
        + "3–4 phòng"
        + "2 phòng + 1 gác"
    - Phù hợp dùng trong bước tiền xử lý dữ liệu (feature engineering)
      cho các cột như `total_bedroom`, `total_restroom`.
    """

    match = re.search(r'\d+', str(text))
    return int(match.group()) if match else None

df['total_bedroom'] = df['total_bedroom'].apply(extract_number)
df['total_restroom'] = df['total_restroom'].apply(extract_number)

In [ ]:
df = df.reset_index(drop=True)

In [ ]:
df

,title,legal_doc,total_bedroom,total_restroom,original_post_date,price_per_m2_million,price,area,ward/district,city/province,is_legal
0,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",Sổ đỏ / Sổ hồng,3,2,2025-10-24,37.80,3.10,82.0,Dĩ An,Bình Dương,1
1,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",Sổ đỏ / Sổ hồng,2,1,2025-10-27,50.00,2.65,53.0,Quận 10,Hồ Chí Minh,1
2,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",Sổ đỏ / Sổ hồng,2,1,2025-10-25,92.71,4.45,48.0,Bình Tân,Hồ Chí Minh,1
3,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",Sổ đỏ / Sổ hồng,2,1,2025-10-25,59.60,5.90,99.0,Thủ Đức,Hồ Chí Minh,0
4,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",Sổ đỏ / Sổ hồng,5,5,2025-10-24,110.94,7.10,64.0,Bình Tân,Hồ Chí Minh,0
...,...,...,...,...,...,...,...,...,...,...,...
20846,"Bán nhà riêng 108 m2 đường Phan Văn Hớn, Tân T...",Sổ đỏ / Sổ hồng,3,2,2025-03-15,14.35,1.55,108.0,Quận 12,Hồ Chí Minh,1
20847,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",Sổ đỏ / Sổ hồng,4,4,2025-03-15,97.50,5.85,60.0,Thủ Đức,Hồ Chí Minh,1
20848,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,Sổ đỏ / Sổ hồng,2,2,2025-03-15,76.43,4.28,56.0,Tân Phú,Hồ Chí Minh,1
20849,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,Sổ đỏ / Sổ hồng,3,3,2025-03-15,238.33,8.58,36.0,Tây Hồ,Hà Nội,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20851 entries, 0 to 20850
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   title                 20851 non-null  object        
 1   legal_doc             20851 non-null  object        
 2   total_bedroom         20851 non-null  int64         
 3   total_restroom        20851 non-null  int64         
 4   original_post_date    20851 non-null  datetime64[ns]
 5   price_per_m2_million  20851 non-null  float64       
 6   price                 20851 non-null  float64       
 7   area                  20851 non-null  float64       
 8   ward/district         20851 non-null  object        
 9   city/province         20851 non-null  object        
 10  is_legal              20851 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(3), object(4)
memory usage: 1.8+ MB


## o. Sắp xếp thứ tự các cột

In [ ]:
new_order = [
    "title",
    "original_post_date",
    "ward/district",
    "city/province",
    "legal_doc",
    "is_legal",
    "area",
    "total_bedroom",
    "total_restroom",
    "price_per_m2_million",
    "price"
]

df = df[new_order]
df

,title,original_post_date,ward/district,city/province,legal_doc,is_legal,area,total_bedroom,total_restroom,price_per_m2_million,price
0,"Bán căn hộ TT AVIO, 82m2, nội thất cơ bản, tha...",2025-10-24,Dĩ An,Bình Dương,Sổ đỏ / Sổ hồng,1,82.0,3,2,37.80,3.10
1,"BÁN CHUNG CƯ CMT8, TRUNG TÂM QUẬN 10, 96 CĂN 5...",2025-10-27,Quận 10,Hồ Chí Minh,Sổ đỏ / Sổ hồng,1,53.0,2,1,50.00,2.65
2,"Bán nhà Tân Hòa Đông, Bình Tân, diện tích 48m²...",2025-10-25,Bình Tân,Hồ Chí Minh,Sổ đỏ / Sổ hồng,1,48.0,2,1,92.71,4.45
3,"BÁN NHÀ HÀNG THƠM, CẦU TĂNG LONG, LÃ XUÂN OAI,...",2025-10-25,Thủ Đức,Hồ Chí Minh,Sổ đỏ / Sổ hồng,0,99.0,2,1,59.60,5.90
4,"Bán nhà phố Phan Anh, Bình Tân, 5 tầng, diện t...",2025-10-24,Bình Tân,Hồ Chí Minh,Sổ đỏ / Sổ hồng,0,64.0,5,5,110.94,7.10
...,...,...,...,...,...,...,...,...,...,...,...
20846,"Bán nhà riêng 108 m2 đường Phan Văn Hớn, Tân T...",2025-03-15,Quận 12,Hồ Chí Minh,Sổ đỏ / Sổ hồng,1,108.0,3,2,14.35,1.55
20847,"Nhà bán trệt 2 lầu ngay Quốc Lộ 13, Phường Hiệ...",2025-03-15,Thủ Đức,Hồ Chí Minh,Sổ đỏ / Sổ hồng,1,60.0,4,4,97.50,5.85
20848,Hxh Tây Thạnh-Tân Phú 56m-2tầng - 4 tỷ 820,2025-03-15,Tân Phú,Hồ Chí Minh,Sổ đỏ / Sổ hồng,1,56.0,2,2,76.43,4.28
20849,Sang nhượng nhà riêng Lạc Long Quân Quận Tây H...,2025-03-15,Tây Hồ,Hà Nội,Sổ đỏ / Sổ hồng,0,36.0,3,3,238.33,8.58


In [ ]:
missing_values = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Mising Values": df.isnull().sum(),
    "Missing Percent": df.isnull().sum() / len(df) * 100
})
missing_values

,Data_Type,Mising Values,Missing Percent
title,object,0,0.0
original_post_date,datetime64[ns],0,0.0
ward/district,object,0,0.0
city/province,object,0,0.0
legal_doc,object,0,0.0
is_legal,int64,0,0.0
area,float64,0,0.0
total_bedroom,int64,0,0.0
total_restroom,int64,0,0.0
price_per_m2_million,float64,0,0.0


# p. Xử lý các cặp giá trị trùng lặp

In [ ]:
df.duplicated().sum()

np.int64(157)

**Nhận xét:** Phát hiện tổng cộng 157 dòng trùng lặp dữ liệu

**Phương án giải quyết:** Loại bỏ các giá trị trùng lặp trên

In [ ]:
df = df.drop_duplicates()
df.duplicated().sum()

np.int64(0)

## q. Lưu dữ liệu đã xử lý

In [ ]:
df.to_csv("/content/drive/MyDrive/DS111_Data/ThuVienNhaDat_Processed.csv")

------------------------------------------------------------ **HẾT** ------------------------------------------------------------